src/main/java/org/ossgang/commons/observables/weak/WeakMethodReferenceObserver.java

```diff
@@ -77,27 +107,44 @@ public void onException(Throwable t) {
     public void onSubscribe(Subscription subscription) {
         synchronized (subscriptions) {
             subscriptions.add(subscription);
-            Optional.ofNullable(holder.get()).ifPresent(h -> subscriptionCountUpdated.accept(h, subscriptions.size()));
+            Optional.ofNullable(holderRef.get()).ifPresent(h -> subscriptionCountUpdated.accept(h, subscriptions.size()));
         }
     }
 
     @Override
     public void onUnsubscribe(Subscription subscription) {
         synchronized (subscriptions) {
             subscriptions.remove(subscription);
-            Optional.ofNullable(holder.get()).ifPresent(h -> subscriptionCountUpdated.accept(h, subscriptions.size()));
+            Optional.ofNullable(holderRef.get()).ifPresent(h -> subscriptionCountUpdated.accept(h, subscriptions.size()));
         }
     }
 
     private <X> void dispatch(BiConsumer<? super C, X> consumer, X item) {
-        C holderInstance = holder.get();
+        C holderInstance = holderRef.get();
         if (holderInstance != null) {
             consumer.accept(holderInstance, item);
         } else {
-            synchronized (subscriptions) {
-                subscriptions.forEach(Subscription::unsubscribe);
-                subscriptions.clear();
+            unsubscribeAll();
+        }
+    }
+
+    private static class WeakCleaner {
+
+        private final WeakReference<?> holderRef;
+        private final Runnable cleanupAction;
+
+        public WeakCleaner(WeakReference<?> holderRef, Runnable cleanupAction) {
+            this.holderRef = holderRef;
+            this.cleanupAction = cleanupAction;
+        }
+
+        public boolean attemptToClean() {
+            if (holderRef.get() == null) {
+                Maybe.attempt(cleanupAction::run);
```

Comment:
This is a little dangerous, if there is an exception for any reason we swallow it ...

I think I would just call `cleanupAction.run()` - if it fails, we get it in the uncaught exception handler

Reply:
I agree to use the `ExceptionHandlers` to at least provide an error, but will have to be done manually. I don't see how it'll be called automatically :slightly_smiling_face: 

I wanted to keep it simple, but probably returning false in case of exception will give the chance of unsubscribing on every cleanup cycle :wink:

Reply:
No, I mean really just not catching the (unchecked) exception at all - in this case it will escape and Java will automatically dispatch it to the default uncaught exception handler?

Reply:
Ahh ok I see, I thought you meant "our" exception handler hahah there is a drawback and an opportunity to that.. let's discuss later in person :wink: